### DISCUSSION TITLE
Using pixel for extra usage?

### DISCUSSION THREAD

**Discussion by frankonski** [HomeSpan/HomeSpan#871] [Q&A] [answered]
I have not seen this officially documented, but when using `setStatusPixel()` with a dev board that sports a neopixel, is there a way for me to use it on top of HomeSpan’s own usage?

Example, assuming I initialize HomeSpan so that the pixel is red, I would know that any time that pixel is red it would report a HomeSpan status.

I would like to be able to set the pixel to, let say, green and have my own blinking patterns for my own purposed, using some kind of “status stack”. I’d push my status to the stack, the pixel would turn green and blink my patterns until I pop out that status to return to HomeSpan’ pixel color: red.

Doable?

**Comment by HomeSpan** [comment]
There is no easy way to override the LED or Pixel used for status once initialized, but you could instead take complete control of the pixel from the outset using `setStatusCallback` to set the pixel either based on status changes, or for whatever other reasons.

You can also try using the `setStatusAutoOff()` method.  This turns off the status LED/Pixel after a period of time.  Once off, you can then control it any way you'd like.  HomeSpan will automatically reset it to the appropriate pattern if the status changes.

**Reply by frankonski** [reply]
Assuming an alarm is triggered on my SecuritySystem, I can’t wait for `statusCallback` to be called by HS..

What if `setStatusPixel()` could receive a reference to my own instance of a Pixel?
.. or just make the Pixel object public/protected?

**Comment by HomeSpan** [comment]
Why would you be waiting?  StatusCallback is only used if HomeSpan needs to set the Pixel.  If you are using it for another purpose, such as an alarm warning, you would set that directly.  The two uses are independent.

That said, big picture is you should be using two LEDs --- anything that tries to dual-purpose the same LED is likely to run into edge cases and not worth changing any of the library code at this point.

**Reply by frankonski** [reply]
Sure I can call that callback myself, but that still does not give me access to the pixel object.

> you would set that directly

That is the thing, the pixel is not publicly available.

> anything that tries to dual-purpose the same LED is likely to run into edge cases

Let me see if I can come up with a draft PR that would please you. The thing is that my project has evolved from an ESP32 dev board on a breadboard using two LEDs (one blue LED for HS, one 3-PIN RGB LED) to a different ESP32-S2 dev board with built-in pixel that is connected via soldered connectors on a prototype board. So by using the built-in pixel of the S2, I am actually saving PINs and don’t need an extra pixel/RGB LED. That makes the proto board “production-ready-enough” for my personal use, and quite leaner.

**Reply by frankonski** [reply]
Oh, and bottom line, the answer to "Using pixel for extra usage?" is currently **No**, correct?

**Reply by HomeSpan** [reply, marked as answer]
You can use it for anything you want - it may or may conflict with what HomeSpan is using it for at the same time.  Once a device is paired to HomeKit the pixel is set to a solid color and that's it.  If you want to use the pixel for something else you can do that, but you will need to reset it back to the solid color as before when you are done.

**Reply by frankonski** [reply]
> If you want to use the pixel for something else you can do that, but you will need to reset it back to the solid color as before when you are done.

oh! wait, I was about to say that my point was that `Blinkable *statusDevice` is private to the `Span` class, but I ended up RTFM a bit more and found `setStatusDevice()`.

So instead of using `setStatusPixel()` I should be able to create my own `Blinkable` device and call `setStatusDevice()` instead, which is what I was referring to when asking _What if setStatusPixel() could receive a reference to my own instance of a Pixel?_

I think this will likely fit my needs.

**Reply by frankonski** [reply]
ok, after fiddling around, I better understand now what you meant by "_it may or may not conflict with what HomeSpan is using it for at the same time._"

After adding a `blinkingColor` to the `Pixel` class, and instantiating a `Blinker *`, everything is fine, I can control the color and blinking, but as soon as HS has a state change, this ends up messing up the state of the Pixel, where I would want it to be.

Lessons learned, don't share the HS status LED/Pixel with another task.

**Comment by MaffooClock** [comment]
For anyone who might land here from Google:

In discussion #1047, the idea has been proposed that the NeoPixel status device _can_ be multi-purpose, but instead of augmenting the blinking behavior, we just specify different colors.
